pkg/apis/azuredisk/v1alpha1/types.go

```diff
@@ -87,26 +85,59 @@ type AzVolumeAttachmentList struct {
 }
 
 // +genclient
-// +genclient:noStatus
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// AzDriverNode is a specification for a AzDriverNode resource
+// AzDriverNode is a representation of a node, where azure CSI driver node plug-in runs.
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="NodeName",type=string,JSONPath=`.spec.nodeName`,description="Name of the Node which this AzDriverNode object represents."
+// +kubebuilder:printcolumn:name="ReadyForVolumeAllocation",type=boolean,JSONPath=`.status.readyForVolumeAllocation`,description="Indicates if the azure persistent volume driver is ready for new pods which use azure persistent volumes."
+// +kubebuilder:printcolumn:name="LastHeatbeatTime",type=integer,JSONPath=`.status.lastHeatbeatTime`,description="Represents the time stamp at which azure persistent volume driver sent a heatbeat."
 type AzDriverNode struct {
 	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec   AzDriverNodeSpec   `json:"spec"`
-	Status AzDriverNodeStatus `json:"status"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec defines the desired state of a AzDriverNode.
+	// Required.
+	Spec   AzDriverNodeSpec   `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status represents the current state of AzDriverNode.
+	// If this is nil or empty, clients should prefer other nodes
+	// for persistent volume allocations or pod places for pods which use azure persistent volumes.
+	// +optional
+	Status *AzDriverNodeStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

Comment:
Should this really be an optional field, or can a client simply rely on a default `Status` value to avoid allocations?

Reply:
I do not have a strong opinion about this, so I'll change if you think we should make it required.

According to documentation, /Status is actually an optional sub-resource, so schema here is in that spirit.

The overall tooling and API semantics also takes /Status as optional, so much so that a regular update won't update the /Status, and UpdateStatus needs to be called for that (this is to make sure we can update status without rev'ing the objection version.).

Reply:
Yes, from the perspective of API sementics I agree that this needs to be optional. However, I do not see any other `Status` fields in the K8s API surface that specify this field by reference rather than by value. Our extenstion should be consistent with the K8s API (and omit the `*`).

Reply:
It's actually based on the recommendation here:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

Also, if you look at the snapshot CRD implementation, they follow the same pattern:
https://github.com/kubernetes-csi/external-snapshotter/blob/release-3.0/client/apis/volumesnapshot/v1beta1/types.go

Reply:
Ah, I see. Thanks for the reference!